01-resources-v2/git-example/artifact

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env ruby
+
+require "json"
+require "git"
+
+# $request = JSON.parse(STDIN.read, symbolize_names: true)
+
+def commit_versions(log)
+  log.collect do |c|
+    {
+      version: {ref: c.sha},
+      metadata: [
+        {name: "author", value: c.author.name},
+        {name: "author_date", value: c.author_date},
+        {name: "commit", value: c.sha},
+        {name: "committer", value: c.committer.name},
+        {name: "committer_date", value: c.committer_date},
+        {name: "message", value: c.message}
+      ]
+    }
+  end
+end
+
+case ARGV[0]
+when "check"
+  $request = {
+    config: {uri: "https://github.com/vito/booklit"},
+    from: {master: {ref: "40bc6986197e411471d306bb8eb3a21c5b5f9d26"}}
```

Comment:
Bit confused here. From [this](https://github.com/vito/rfcs/blame/217f5754433de8a702e46c1fb8a90621bb6c7816/01-resources-v2/proposal.md#L54) I understand that `check` should receive a `Space` and a `Version`. And [`Version`](https://github.com/vito/rfcs/blame/217f5754433de8a702e46c1fb8a90621bb6c7816/01-resources-v2/proposal.md#L42) should be a dict ( e.g. `{ "ref": "" }`), while [`Space`](https://github.com/vito/rfcs/blame/217f5754433de8a702e46c1fb8a90621bb6c7816/01-resources-v2/proposal.md#L39) should just be a string?

Based on the above, I'd expect to see something like this:

```json
{
  "space": "master",
  "version": { "ref": "40bc6986197e411471d306bb8eb3a21c5b5f9d26" }
}
```

Reply:
I took a liberty here and used the space as the key in the map, instead of having a list of JSON objects like that. It makes it easier to find the 'from' version when checking for each branch.

This is really the only reason I switched from JSON objects to strings for the space representation.

Reply:
Correct me if I'm wrong - but wouldn't Go code look like this then:

```go
type CheckRequest struct {
  config  Config `json:"config"`
  version map[string]map[string]string `json:"from"`
}
```

If so, that seems less "clean" than having:
```go
type CheckRequest struct {
  config  Config `json:"config"`
  version Version `json:"version"`
}

type Version struct {
  Space string `json:"space"` // Or map[string]string
  Version map[string]string `json:"version"`
}
```

Reply:
To be fair, you've got a `map[string]string` there where really the resource would have its own type. So it'd be:

```go
type Version struct {
  Ref string `json:"ref"`
}

type CheckRequest struct {
  Config Config `json:"config"`
  From map[string]Version `json:"from"`
}
```

It may not be as pretty on the type definition, but it's definitely prettier when it comes to finding the version to check from:

```go
for _, branch := range branches {
  var version Version
  for _, f := range request.From {
    if f.Space == branch {
      version = f.Version
    }
  }

  // git log {version.Ref}..HEAD
}
```

vs.

```go
for _, branch := range branches {
  version := request.From[branch]
  // git log {version.Ref}..HEAD
}
```

Of course, you could clean it up with a helper method, but it's still more code that the author has to write.

Reply:
I see what you mean, and it does look better:) Just have one question; `check` requests will contain the latest version for all discovered spaces? (And not just the latest across spaces?). In your Ruby example it's just a single space and version, figure it might be worth updating the example to reflect what will actually be passed to the resource.

Reply:
Yep. I'll update the example to be more clear.